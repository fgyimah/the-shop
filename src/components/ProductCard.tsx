import React from 'react';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { Product } from '../@types';
import { RootState } from '../store/root';
import { addItemToCart } from '../store/cart.reducer';

interface Props {
    product: Product;
}

const ProductCard: React.FC<Props> = ({ product }) => {
    const history = useHistory();
    const cart = useSelector((state: RootState) => state.cart);
    const dispatch = useDispatch();

    const addedToCart = () => {
        return cart.items.findIndex((item) => item.product.id === product.id) !== -1;
    };

    const addToCart = () => {
        if (addedToCart()) return;
        dispatch(addItemToCart(product));
    };

    return (
        <StyledCard>
            <div className="image" onClick={() => history.push(`/products/${product.id}`)}>
                <img src={product.imageUrl} alt={product.name} data-testid="product-image" />
            </div>
            <div className="info" onClick={() => history.push(`/products/${product.id}`)}>
                <h4 data-testid="product-name">{product.name}</h4>
                <p className="item-description" data-testid="product-description">
                    {product.description}
                </p>
            </div>
            <div
                className={`add-to-cart-btn ${addedToCart() ? 'added-to-cart' : ''}`}
                onClick={addToCart}
            >
                {addedToCart() ? 'ADDED TO CART' : 'ADD TO CART'}
            </div>
            <div
                className="price-div"
                onClick={() => history.push(`/products/${product.id}`)}
                data-testid="product-price"
            >
                GHâ‚µ {product.price}
            </div>
        </StyledCard>
    );
};

const StyledCard = styled.div`
    box-shadow: 0 16px 16px -2px rgba(0, 0, 0, 0.4);
    margin: 1.5rem;
    height: max-content;
    position: relative;
    cursor: pointer;

    .image {
        img {
            width: 100%;
            height: 22rem;
        }
    }
    .info {
        text-align: center;
    }
    .add-to-cart-btn {
        background-color: #000;
        margin-top: 6px;
        color: #fff;
        padding: 1rem 2rem;
        cursor: pointer;
        width: 100%;
        text-align: center;
    }
    .added-to-cart {
        background-color: #2c2929;
        cursor: not-allowed;
    }

    .item-description {
        color: #3d3b3b;
    }

    .price-div {
        position: absolute;
        top: 10px;
        right: 10px;
        background-color: #000;
        background-size: cover;
        padding: 2rem;
        clip-path: polygon(0% 20%, 60% 20%, 60% 0%, 100% 50%, 60% 100%, 60% 80%, 0% 80%);
        animation: clips 3s ease infinite;
        color: #fff;
    }

    @keyframes clips {
        50% {
            clip-path: polygon(40% 0%, 40% 20%, 100% 20%, 100% 80%, 40% 80%, 40% 100%, 0% 50%);
        }
    }
`;

export default ProductCard;
