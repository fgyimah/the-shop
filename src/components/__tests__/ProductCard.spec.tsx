import { render } from '@testing-library/react';
import * as reactRedux from 'react-redux';
import ProductCard from '../ProductCard';

jest.mock('react-redux', () => ({
    ...(jest.requireActual('react-redux') as any),
    useSelector: jest.fn(),
}));

describe('Component Test - ProductCard', () => {
    const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
    const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

    beforeEach(() => {
        useSelectorMock.mockClear();
        useDispatchMock.mockClear();
    });

    it('matches snapshot', () => {
        useSelectorMock.mockReturnValue({ items: [] });
        const data = {
            product: {
                id: '1',
                name: 'fake-product',
                description: 'fake description',
                price: 99,
                imageUrl: '',
            },
        };
        const { asFragment } = render(<ProductCard {...data} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders using data passed to it', () => {
        useSelectorMock.mockReturnValue({ items: [] });
        const data = {
            product: {
                id: '1',
                name: 'fake-product',
                description: 'fake description',
                price: 99,
                imageUrl: '',
            },
        };
        const { getByTestId } = render(<ProductCard {...data} />);
        expect(getByTestId('product-name').innerHTML).toMatch(data.product.name);
        expect(getByTestId('product-description').innerHTML).toMatch(data.product.description);
        expect(getByTestId('product-price').innerHTML).toMatch(`GHâ‚µ ${data.product.price}`);
    });
});
