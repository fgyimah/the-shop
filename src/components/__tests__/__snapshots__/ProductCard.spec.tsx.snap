// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Test - ProductCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ cRUtw"
  >
    <div
      class="image"
    >
      <img
        alt="fake-product"
        data-testid="product-image"
        src=""
      />
    </div>
    <div
      class="info"
    >
      <h4
        data-testid="product-name"
      >
        fake-product
      </h4>
      <p
        class="item-description"
        data-testid="product-description"
      >
        fake description
      </p>
    </div>
    <div
      class="add-to-cart-btn "
    >
      ADD TO CART
    </div>
    <div
      class="price-div"
      data-testid="product-price"
    >
      GHâ‚µ 99
    </div>
  </div>
</DocumentFragment>
`;
